gensim/models/_fasttext_bin.py

```diff
@@ -197,6 +194,41 @@ def _load_vocab(fin, new_format, encoding='utf-8'):
     return raw_vocab, vocab_size, nwords
 
 
+def _decode_bad_unicode_py2(word_bytes, encoding):
+    #
+    # We are explicitly ignoring the encoding here.  This is bad, because the
+    # output is highly likely to be garbage, but its the only thing that:
+    #
+    # - does not cause collisions (see https://github.com/RaRe-Technologies/gensim/issues/2402)
+    # - works under Py2
+    #
+    del encoding
+    word = word_bytes.decode('latin1')
+    logger.error(
+        'failed to decode invalid unicode bytes %r; falling back to latin1, using %r; '
+        'consider upgrading to Python 3 for better Unicode handling (and so much more)',
```

Comment:
Seems unrelated. The problem is with badly encoded strings in the model (FB's fault), or possibly our choice to accept such inputs instead of failing explicitly.

There's nothing wrong with Python 2, and Python 3 won't fix the problem (just mask it in a different way).

Reply:
Under Py2, you get a string of complete garbage. The term is essentially lost from the vocabulary. Under Py3, you get a mostly OK string with the broken characters escaped. The term remains in the vocabulary, and can contribute vectors to the model. See the unit tests for an example.

So under Py2, our handling of this edge case is worse than under Py3.

Reply:
Both terms are garbage and lost from the vocabulary, I see no difference. No user query or input text will ever match them, whether it's latin1-garbage or backslash-escaped-garbage.

The only useful property might be *substrings* that still happen to be decoded correctly within that term (probably ~ASCII subset; true both for latin1 and backslash-escaped).

IMO the correct solution would be to "refuse to guess" and either:
1. fail the model loading with an exception, in case valid unicode (via utf8?) is the contract of the FT format, aka "not our problem"; or
2. work with bytes instead of text, in case there's no decoding contract (which I hope is not the case, because that's a lot of work, fixing all the interfaces that expect text!).

Which approach does the Facebook's reference implementation use?

Either way, if we want to go down the road of silently modifying the models, the difference of py2 vs py3 is moot. Anything you can do with unicode in python3, you can also do in python2. We'd just implement `backslashreplace` (looks easy enough) and be consistent, instead of muddying the water by logging arcane py2/3 upgrade recommendations.

Reply:
>  We'd just implement backslashreplace (looks relatively easy) 

[Googling](https://stackoverflow.com/questions/42860186/exact-equivalent-of-b-decodeutf-8-backslashreplace-in-python-2) turned out this, as a possible `backslashreplace` backport:

```python
def backslashescape_backport(ex):
    # The error handler receives the UnicodeDecodeError, which contains arguments of the
    # string and start/end indexes of the bad portion.
    bstr, start, end = ex.object, ex.start, ex.end

    # The return value is a tuple of Unicode string and the index to continue conversion.
    # Note: iterating byte strings returns int on 3.x but str on 2.x
    return u''.join('\\x{:02x}'.format(c if isinstance(c, int) else ord(c)) for c in bstr[start:end]), end

codecs.register_error("backslashescape_backport", backslashescape_backport)
bad_term = u"Žluťoučký koníček".encode("latin2")  # create invalid UTF8
print(repr(bad_term.decode("utf-8", "backslashescape_backport")))
u'\\xaelu\\xbbou\\xe8k\\xfd kon\\xed\\xe8ek'
```